File: algorithms/medium/check_if_a_string_can_break_another_string.go
package main

import (
        "fmt"
        "sort"
)

func checkIfCanBreak(s1 string, s2 string) bool {
    arr1 := []int{}
    for _, v := range s1 {
        arr1 = append(arr1, int(v))
    }
    sort.Ints(arr1)
    arr2 := []int{}
    for _, v := range s2 {
        arr2 = append(arr2, int(v))
    }
    sort.Ints(arr2)
    cb1, cb2 := true, true
    for i := 0; i < len(s1); i++ {
        cb1 = cb1 && arr1[i] >= arr2[i]
        cb2 = cb2 && arr2[i] >= arr1[i]
    }
    return cb1 || cb2
}

func main() {
    data := [][]string{
                []string{"abc", "xya"},
                []string{"abe", "acd"},
                []string{"leetcodee", "interview"},
                []string{"mumesxoosf", "krlwrpzneb"},
                []string{"lgwipivkjpnqjncxuaaeiqhlikoypbrilouwqnegkkhurtccwfeeqdttcxlojzvltczuawpjkpudoxvtjqdaxtjinjkcxmlupyfa", "gqelahfdvwilykowkvpqgotcerhhfdsgxvzzwqqmauhanzuwpymgstidpptazyflcnixasesuavqftjncdcwafzgbubsbwrwpzmm"},
                []string{"zrxjnkocvoqkwysqcclzbqmaojngaiohizahztfyhctarpnobizdvgainnzrlifalfpbgzzcqcpjdauwjnqfpwuteapivythljuosiwwhlecouxficbwqlcvbqudnabhxjisppzepyjfxcnvqhmfhcxewdmwpzboadzfzhrxhcsxvnxwttwquiaukpstiorecjitrqxrvusobyywolzcdnswxtkfcwivvymrywkjnwhivnmaxnzduegvzjsvhdgwgpvjdngmjympwssilqhtpfpuxcbbmqfstddlqyaduwshwoctjqawxyvqwyrfojghxvfqahkrqjtzwcymhhcptcnpuursyfiblezevrecjhjcuwazdnmmamvvbywxmldplehozxgvgvzzdspolorehymztmaxvtqkotassdgkpkjazeskmflemebmodohtumnvbsbczquupyvpvpyorhluleadlxugduslqpitlhvkiuvxmmkqvvgpwghgrkiyniutkyvpuvxqpphgsqqvuwojuyptkvriyhyjukkaipnjlqkenlapzzaqgzbsxdgkklibgoyqqzuadlzjzyljkqcwqqmtxcwpsicqmcjvmfdwgdlqcmbmhgpnmxweojfbmqidfbrjwjahbeizyrtihvwalifizrwpcfgsjytpiipmxyldkuhnukdwyfsscanunfzisfnbadbenkdllxwkvsumkxjmkqqlxdeahwngeqivnsbefeuvqevinjotbruaterhwaciqltbyoipkpivklchyxjsstasymjlcnmvyzktcoowrnyubvmcnvclpktxfwcnsovoyhgjvhrbzyrmiotykqhhhhdcnuwuyxwaomqsfiadlzlmqxkokrcpyckimameihjejkeqpizzdkhsukdsxqazvoarjkhogrpejpugpexfrjfazmvryzuqydtmorwrwgoqqofspdsmmgigvlihoyxaeuwbkvkcdvfqpxm", "ljbhqpixqarpxwlfdbamrnjbhrtxjtyrwntdatpudvhcdwlqjmtsvedemkqkanvewnruzqqegaxzagbnktqwrxgdbopobdyflkevfavnghhygweljmptvtbintxllvqzpwhjsienrpxctyyruikwjcxmmothscuowmpavytbxnknkuqsivioncmpiwbonuvoorwmzzkvqlkagaimrhbqmxumsvpqlzglpytpxeglkeuzocgtriterqypjkkogujuhzjpkbtopiixgbcciptevoncylihbjjkwftzrerwzatkedlpespelaxhdfngadboxzgsozwlmkdzgttiaaitvkhqamshvxpohnovyycbfdjleaclllwnbaezjsemynfjfffqvuyduboouxmagryhghklvnbxugyabjanvcyraqysqmtxxgfhlnjexiowxcefxdeopvzlzcwxwvnjytnbzhtsamdvvnevuqzjhisrikklxkhnnvvtfprkiqagmjoqzrlobgpclhwxfzoewsuyudidlqgpcfgropyraowysiyvpsylabcujcwpfmpznmprpqqlircfndeegmxtcfsrhqvapgbxwlwtjmnznllxzwoyyoltpezhlsoqlidhdrooxjaazfzzbvwrsmwrahpevcqqamobpysjhwvwxhgprrechefoxaeenjpcfqzdhjwhcgnjouirmlualziygmzakldgdffyxpzlahabtwynbpuhjdgyateevqbrketeusxtamqahiqdxxtvoqwsqqtymuvblkoleunqnomrhaovynjhffhvxosoqftnoihieocxirccwmenqhyyxahepmoqzuqegcjrbwfysjanwjjcjrlwmpmhbkhtdyevbpquuxwrldbqzbmjnvfkhynkoekccdspieznajtplvinmnneufikadohiwgaspvzfhljvzswbnpnbpnwomqhxsnnfckpfdizoirjavsujdxfzlhy"},
            }
    for _, v := range data {
        s1, s2 := v[0], v[1]
        fmt.Printf("s1, s2 = %s, %s\n", s1, s2)
        r := checkIfCanBreak(s1, s2)
        fmt.Printf("r = %v\n", r)
        fmt.Println("=================")
    }
}

File: algorithms/medium/check_if_a_string_can_break_another_string.py
class Solution:
    def checkIfCanBreak(self, s1: str, s2: str) -> bool:
        '''
            s1 = "leetcodee" ; "c d eeee    l   o   t"
            s2 = "interview" ; "    ee   ii   n   r t v w"
            s1 = "c d ee    l   o"
            s2 = "       ii   n   r v w"
        '''
        s1_sorted = ''.join(sorted(s1))
        s2_sorted = ''.join(sorted(s2))
        #can_break = True
        #for x, y in zip(s1_sorted, s2_sorted):
        #    if x < y:
        #        can_break = False
        #if can_break:
        #    return True
        #can_break = True
        #for x, y in zip(s2_sorted, s1_sorted):
        #    if x < y:
        #        can_break = False
        #return can_break
        return all([x >= y for x, y in zip(s1_sorted, s2_sorted)]) or all([x >= y for x, y in zip(s2_sorted, s1_sorted)])

# Main section
for s1, s2 in [
                 ('abc', 'xya'),
                 ('abe', 'acd'),
                 ('leetcodee', 'interview'),
                 ('mumesxoosf', 'krlwrpzneb'),
                 ('lgwipivkjpnqjncxuaaeiqhlikoypbrilouwqnegkkhurtccwfeeqdttcxlojzvltczuawpjkpudoxvtjqdaxtjinjkcxmlupyfa', 'gqelahfdvwilykowkvpqgotcerhhfdsgxvzzwqqmauhanzuwpymgstidpptazyflcnixasesuavqftjncdcwafzgbubsbwrwpzmm'),
                 ('zrxjnkocvoqkwysqcclzbqmaojngaiohizahztfyhctarpnobizdvgainnzrlifalfpbgzzcqcpjdauwjnqfpwuteapivythljuosiwwhlecouxficbwqlcvbqudnabhxjisppzepyjfxcnvqhmfhcxewdmwpzboadzfzhrxhcsxvnxwttwquiaukpstiorecjitrqxrvusobyywolzcdnswxtkfcwivvymrywkjnwhivnmaxnzduegvzjsvhdgwgpvjdngmjympwssilqhtpfpuxcbbmqfstddlqyaduwshwoctjqawxyvqwyrfojghxvfqahkrqjtzwcymhhcptcnpuursyfiblezevrecjhjcuwazdnmmamvvbywxmldplehozxgvgvzzdspolorehymztmaxvtqkotassdgkpkjazeskmflemebmodohtumnvbsbczquupyvpvpyorhluleadlxugduslqpitlhvkiuvxmmkqvvgpwghgrkiyniutkyvpuvxqpphgsqqvuwojuyptkvriyhyjukkaipnjlqkenlapzzaqgzbsxdgkklibgoyqqzuadlzjzyljkqcwqqmtxcwpsicqmcjvmfdwgdlqcmbmhgpnmxweojfbmqidfbrjwjahbeizyrtihvwalifizrwpcfgsjytpiipmxyldkuhnukdwyfsscanunfzisfnbadbenkdllxwkvsumkxjmkqqlxdeahwngeqivnsbefeuvqevinjotbruaterhwaciqltbyoipkpivklchyxjsstasymjlcnmvyzktcoowrnyubvmcnvclpktxfwcnsovoyhgjvhrbzyrmiotykqhhhhdcnuwuyxwaomqsfiadlzlmqxkokrcpyckimameihjejkeqpizzdkhsukdsxqazvoarjkhogrpejpugpexfrjfazmvryzuqydtmorwrwgoqqofspdsmmgigvlihoyxaeuwbkvkcdvfqpxm', 'ljbhqpixqarpxwlfdbamrnjbhrtxjtyrwntdatpudvhcdwlqjmtsvedemkqkanvewnruzqqegaxzagbnktqwrxgdbopobdyflkevfavnghhygweljmptvtbintxllvqzpwhjsienrpxctyyruikwjcxmmothscuowmpavytbxnknkuqsivioncmpiwbonuvoorwmzzkvqlkagaimrhbqmxumsvpqlzglpytpxeglkeuzocgtriterqypjkkogujuhzjpkbtopiixgbcciptevoncylihbjjkwftzrerwzatkedlpespelaxhdfngadboxzgsozwlmkdzgttiaaitvkhqamshvxpohnovyycbfdjleaclllwnbaezjsemynfjfffqvuyduboouxmagryhghklvnbxugyabjanvcyraqysqmtxxgfhlnjexiowxcefxdeopvzlzcwxwvnjytnbzhtsamdvvnevuqzjhisrikklxkhnnvvtfprkiqagmjoqzrlobgpclhwxfzoewsuyudidlqgpcfgropyraowysiyvpsylabcujcwpfmpznmprpqqlircfndeegmxtcfsrhqvapgbxwlwtjmnznllxzwoyyoltpezhlsoqlidhdrooxjaazfzzbvwrsmwrahpevcqqamobpysjhwvwxhgprrechefoxaeenjpcfqzdhjwhcgnjouirmlualziygmzakldgdffyxpzlahabtwynbpuhjdgyateevqbrketeusxtamqahiqdxxtvoqwsqqtymuvblkoleunqnomrhaovynjhffhvxosoqftnoihieocxirccwmenqhyyxahepmoqzuqegcjrbwfysjanwjjcjrlwmpmhbkhtdyevbpquuxwrldbqzbmjnvfkhynkoekccdspieznajtplvinmnneufikadohiwgaspvzfhljvzswbnpnbpnwomqhxsnnfckpfdizoirjavsujdxfzlhy'),
              ]:
    print(f's1, s2 = {s1}, {s2}')
    sol = Solution()
    r = sol.checkIfCanBreak(s1, s2)
    print(f'r = {r}')
    print('==================')


